### PR TITLE
Simplify Intersection of Sets involving FiniteSet

### LINKED ISSUE #9623: Intersection(S.Reals, Interval(0, oo), FiniteSet(n)) is not simplified
```python
In [10]: Intersection(S.Reals, Interval(0, oo), FiniteSet(n))
Out[10]: Intersection((-oo, oo), [0, oo), {n})
```

It should simplify to the following:
```python
In [10]: Intersection(S.Reals, Interval(0, oo), FiniteSet(n))
Out[10]: Intersection([0, oo), {n})
```

### PR DESCRIPTION
**Pull request by aktech**
- [x] Fixes #9623 

Add Pairwise rule for Intersection of Sets involving S.Reals.

```
In [10]: Intersection(S.Reals, Interval(0, oo), FiniteSet(n))
Out[10]: Intersection([0, oo), {n})
```

### CODE REVIEW

**Comment by aktech** [comment]
@hargup  @flacjacket

**Comment by leosartaj** [comment]
This maybe done by implementing ``_intersect`` in ``Reals`` just like other sets?

File: sympy/sets/sets.py [hunk 1]
```diff
@@ -1429,6 +1429,19 @@ def reduce(args):
...
+        # smaller real subset
+
+        def _fsreal(fs):
+            """Check if all element of FiniteSet is real"""
+            if fs.is_FiniteSet:
+                return all(element.is_real for element in fs)
+
+        if S.Reals in args:
+            other_args = [a for a in args if a != S.Reals]
+            if True in [(b.is_Interval or _fsreal(b)) for b in other_args]:
```
Comment (hargup) [inline sympy/sets/sets.py]: You can also use `any` [#1]

[#1]: https://docs.python.org/2/library/functions.html#any

**Comment by gxyd** [comment]
I am not sure if my comments matter, but I think the problem is not with the `S.Reals`  for ex. 
```
>>> Intersction(Interval(1,2), Interval(3,4), FiniteSet(n))
# returned as it is instead of EmptySet
```
As I see the code you are only apllying it to `S.Reals` so I guess my example would not be taken care of.

I think problem is in line no 1442 on sets.py, since if it finds any FiniteSet containing an element which is `None` for its containment. We should also make the `Intersection(star_here other_args)` there (I cannot find star on.my keypad) . This code may work on line 1442.
```
 other_args = list([intersection(star_here other_args)])
If s.emptyset in other_args:
     return s.emptyset
```
Please bare with me if you find anything wrong here.

**Comment by aktech** [comment]
@leosartaj 

> This maybe done by implementing _intersect in Reals just like other sets?

These are Global rules & all global rules should reside at one [place](https://github.com/sympy/sympy/blob/master/sympy/sets/sets.py#L1427).

**Comment by aktech** [comment]
@gxyd Thanks for the review. 

> I am not sure if my comments matter, but I think the problem is not with the S.Reals for ex.

Every comment matters, feel free to comment.

> As I see the code you are only apllying it to S.Reals so I guess my example would not be taken care of.

I see I am only fixing the subset of the problem you are talking about, So my fix isn't general, which is not apt. +1 for your code, it looks good.

> (I cannot find star on.my keypad)

Press Shift + 8

**Comment by aktech** [comment]
@gxyd  @hargup  @leosartaj 

It's up for review now, please have a look.

**Comment by gxyd** [comment]
+1 looks good to me. Thanks for the clarification.

**Comment by leosartaj** [comment]
While you are it. Maybe this can also be fixed.

In both master and your branch
```
>>> Intersection(FiniteSet(m, n), FiniteSet(m))
{m}
```

In master
```
>>> Intersection(S.Reals, Interval(0, oo), FiniteSet(m), FiniteSet(n))
ℝ ∩ [0, ∞) ∩ {m} ∩ {m, n}
```
In your branch
```
>>> Intersection(S.Reals, Interval(0, oo), FiniteSet(m), FiniteSet(n))
[0, ∞) ∩ {m} ∩ {m, n}
```

Problem with ``S.Reals`` is fixed. :+1:

I guess this also can be simplified to give something like this.

```
>>> Intersection(S.Reals, Interval(0, oo), FiniteSet(m), FiniteSet(n))
[0, ∞) ∩ {m}
```

**Comment by aktech** [comment]
@leosartaj  It already works, probably you have not fetched the latest version of my branch.
```python
In [13]: Intersection(S.Reals, Interval(0, oo), FiniteSet(m), FiniteSet(n))
Out[13]: Intersection([0, oo), {m}, {n})
```

**Comment by aktech** [comment]
>I guess this also can be simplified to give something like this.
>
```python
Intersection(S.Reals, Interval(0, oo), FiniteSet(m), FiniteSet(n))
[0, ∞) ∩ {m}
```

Why do you think so?

**Comment by leosartaj** [comment]
I am sorry, a lot of typo these days.

I meant this
```
Intersection(S.Reals, Interval(0, oo), FiniteSet(m, n), FiniteSet(n))
[0, ∞) ∩ {m}
```

**Comment by aktech** [comment]
@leosartaj 
As of now I don't see a clean way to do so, may be I will open an another PR for that in future.

**Comment by gxyd** [comment]
Is it just me who feels that `reduce` in `Intersection` is messed up compared to the `reduce` in `Union` ?

**Comment by aktech** [comment]
@leosartaj  @gxyd  any objection for [this](https://github.com/sympy/sympy/pull/9627#issuecomment-119441143).

**Comment by leosartaj** [comment]
> Is it just me who feels that reduce in Intersection is messed up compared to the reduce in Union ?

I think so too @gxyd 

> any objection for this.

``reduce`` can be cleaned. Maybe a  separate PR can handle that. Other than that +1.

**Comment by gxyd** [comment]
I am +1 too. Probably next week( currently some technical diff with my laptop) I too will try to make `reduce` a little better.

**Comment by aktech** [comment]
>I too will try to make reduce a little better.

Yeah sure, go ahead!

**Comment by hargup** [comment]
+1

**Comment by aktech** [comment]
Thanks @hargup  @gxyd  @leosartaj  for reviewing this.

**Pull request merged**